### PR TITLE
Update socket2 to v0.4

### PR DESCRIPTION
**Pull request by Thomasdezeeuw**

### CODE REVIEW

File: src/adapter.rs [hunk 1]
```diff
@@ -240,16 +240,20 @@ unsafe fn get_adapter(adapter_addresses_ptr: PIP_ADAPTER_ADDRESSES) -> Result<Ad
...
-    );
-
-    // Could be either ipv4 or ipv6
-    sockaddr
-        .as_inet()
-        .map(|s| IpAddr::V4(*s.ip()))
-        .unwrap_or_else(|| IpAddr::V6(*sockaddr.as_inet6().unwrap().ip()))
+    let (_, sockaddr) = socket2::SockAddr::init(|storage, length| {
+        debug_assert!(
+            std::mem::size_of_val(unsafe { &*socket_address.lpSockaddr })
```
Comment (liranringel) [inline src/adapter.rs:245]: This function is marked as `unsafe` so the `unsafe` here is redundant (also in other places).

**Review by liranringel — COMMENTED** [review]
Thanks for the PR!
I had 1 comment about the unsafe.

File: src/adapter.rs (hunk 1 above)
Comment (Thomasdezeeuw) [reply src/adapter.rs]: Done.

**Comment by Thomasdezeeuw** [comment]
@liranringel any chance you have some time to take another look at this?

File: src/adapter.rs [hunk 2]
```diff
@@ -240,16 +240,17 @@ unsafe fn get_adapter(adapter_addresses_ptr: PIP_ADAPTER_ADDRESSES) -> Result<Ad
...
-    // Could be either ipv4 or ipv6
-    sockaddr
-        .as_inet()
-        .map(|s| IpAddr::V4(*s.ip()))
-        .unwrap_or_else(|| IpAddr::V6(*sockaddr.as_inet6().unwrap().ip()))
+    let (_, sockaddr) = socket2::SockAddr::init(|storage, length| {
+        debug_assert!(
+            std::mem::size_of_val(&*socket_address.lpSockaddr) <= std::mem::size_of_val(&*storage)
+        );
+        storage.copy_from_nonoverlapping(socket_address.lpSockaddr as *const _, 1);
```
Comment (liranringel) [inline src/adapter.rs:247]: I think this line will copy only 16 bytes (the size of `SOCKADDR`) but it won't work if the socket is ipv6.
I think exactly `socket_address .iSockaddrLength` bytes should be copied here. Am I right?

**Review by liranringel — COMMENTED** [review]
I've looked at it again and I think there's a problem that I mentioned in a comment.

File: src/adapter.rs (hunk 2 above)
Comment (Thomasdezeeuw) [reply src/adapter.rs]: I've changed it to copy `socket_address.iSockaddrLength` bytes.

**Comment by vorot93** [comment]
@liranringel could you please take a look again? 🙂

**Comment by conblem** [comment]
Any news here @liranringel

**Comment by liranringel** [comment]
@vorot93 @conblem I'd like to add it but it does not compile.
You can see that in the ci: https://ci.appveyor.com/project/liran-ringel/ipconfig/builds/38627628/job/071q8dsg4wjt74d4

**Comment by conblem** [comment]
Thanks for the update i will try to take a look at it

**Comment by conblem** [comment]
I have created #41 which compiles under windows

**Pull request closed without merging**